src/main/java/frc/robot/commands/ShooterCommand.java

```diff
@@ -0,0 +1,99 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2020 1745 JJ Pearce Robotics. All Rights Reserved.           */
+/*----------------------------------------------------------------------------*/
+
+package frc.robot.commands;
+
+import edu.wpi.first.wpilibj.Timer;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.subsystems.HopperController;
+import frc.robot.subsystems.IndexerController;
+import frc.robot.subsystems.shooter.Shooter;
+import frc.robot.subsystems.shooter.ShooterSpeedController;
+import io.github.oblarg.oblog.annotations.Config;
+
+public class ShooterCommand extends CommandBase {
+
+    private Shooter shooter;
+    private ShooterSpeedController shooterSpeedController;
+    private HopperController hopperController;
+    private IndexerController indexerController;
+
+    private final double BALL_SHOOT_TIME = 1.0;
+    private final double INNER_DISTANCE_FROM_TARGET = 29.0;
+
+    private double distanceToGoal = 123.0; // TODO set back to 0 when distance is working
+    private double ballShootStartTime;
+    private boolean shotRequested = false;
+    private boolean shotInProgress = false;
+
+    // Default Constructor. Set initial launch speed to 0 inches/second.
+    public ShooterCommand(Shooter shooter, ShooterSpeedController shooterSpeedController,
+            HopperController hopperController, IndexerController indexerController) {
+        this.shooter = shooter;
+        this.shooterSpeedController = shooterSpeedController;
+        this.hopperController = hopperController;
+        this.indexerController = indexerController;
+    }
+
+    // Called just before this Command runs the first time
+    @Override
+    public void initialize() {
+        // call shooter.determineLaunchSpeed
+        // use it to set the shooterSpeedController
+        // TODO Limelight might take inner distance to account, revisit this
+        shooterSpeedController
+                .setLaunchSpeed(shooter.determineLaunchSpeed(distanceToGoal + INNER_DISTANCE_FROM_TARGET));
+
+        // Set the bool to know that a shot is requested
+        shotRequested = true;
+    }
+
+    // Called repeatedly when this Command is scheduled to run
+    @Override
+    public void execute() {
+
+        // Make sure the shooter is at speed before loading a power cell
+        if (shooterSpeedController.isAtSpeed() && !shotInProgress) {
```

Comment:
What do you think about running the indexer and hopper while the shooter is at speed and not running them otherwise? Then when the command is interrupted (by driving, for example), `end()` is called, where you can turn off the indexer and hopper.

Reply:
How will the teleop command interrupt the other commands?  Should it call cancel on all of them, or should we turn all of our subsystems into WPILib subsystemBases, and have them register themselves, so it should happen automatically?  I vote the subsystemBase way, and we start with Drive.  However, I also think it should be a new feature, so that we can get this code merged to master.

Reply:
Also I will make the change you suggest, that way calling end turns things off.

Reply:
I like the SubsystemBase way. I shouldn't have removed that stuff from the gyroscope PR. I'll make sure to add it back in at tonights meeting.

Reply:
Let's just get the gyro PR merged as is.  We can make a new feature for converting some subsystems to SubsystemBases.  We should start with Drive, so that ShooterCommand can get interrupted by driving (also ShooterCommand will need to use Drive).  Maybe we should make TeleopCommand uninterruptable so that shooting doesn't stop it?

  Also, I broke the ShooterCommand class by editing it (using GitHub, not at an IDE), and I can't figure out why it's not compiling.  I'm guessing I missed a bracket or a parenthesis, could you possibly check it out?  Please feel free to fix it.  If not, I'll fix it first thing tonight.

Reply:
Gracias

Reply:
The ShooterCommand needed to be imported in IO, it's fixed now. 
Re: new feature for converting subsystems, that sounds good